File: snowpipe-streaming-python/src/snowpipe_streaming_manager.py
import logging
from typing import List, Dict, Any, Optional
from snowflake.ingest.streaming import StreamingIngestClient, StreamingIngestChannel
from models import Order, OrderItem
from config_manager import ConfigManager
import snowflake.connector
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.backends import default_backend

logger = logging.getLogger(__name__)


class SnowpipeStreamingManager:
    def __init__(self, config: ConfigManager, instance_id: int = -1):
        self.config = config
        self.instance_id = instance_id
        
        channel_suffix = f"_instance_{instance_id}" if instance_id >= 0 else ""
        logger.info(
            f"Creating Snowflake Streaming clients and opening channels"
            f"{' for instance ' + str(instance_id) if instance_id >= 0 else '...'}"
        )
        
        self.properties = {
            "account": config.get_snowflake_account(),
            "user": config.get_snowflake_user(),
            "private_key": config.get_private_key(),
            "url": config.get_snowflake_url(),
            "role": "SNOWFLAKE_INTELLIGENCE_ADMIN",
            "warehouse": config.get_warehouse(),
        }
        
        self.orders_client = StreamingIngestClient(
            client_name=f"ORDERS_CLIENT_{instance_id}",
            db_name=config.get_database(),
            schema_name=config.get_schema(),
            pipe_name=config.get_property("pipe.orders.name"),
            properties=self.properties,
        )
        
        self.order_items_client = StreamingIngestClient(
            client_name=f"ORDER_ITEMS_CLIENT_{instance_id}",
            db_name=config.get_database(),
            schema_name=config.get_schema(),
            pipe_name=config.get_property("pipe.order_items.name"),
            properties=self.properties,
        )
        
        self.orders_channel = self._open_channel(
            self.orders_client,
            config.get_property("channel.orders.name") + channel_suffix,
        )
        
        self.order_items_channel = self._open_channel(
            self.order_items_client,
            config.get_property("channel.order_items.name") + channel_suffix,
        )
        
        logger.info("All clients and channels initialized successfully")

    def _open_channel(
        self, client: StreamingIngestClient, channel_name: str
    ) -> StreamingIngestChannel:
        initial_offset = "0"
        channel, status = client.open_channel(channel_name, initial_offset)
        
        latest_offset = channel.get_latest_committed_offset_token()
        logger.info(
            f"Channel {channel_name} opened. Latest committed offset: "
            f"{latest_offset if latest_offset else 'NULL (new channel)'}"
        )
        
        return channel

    def get_max_customer_id(self) -> int:
        private_key_pem = self.config.get_private_key()
        private_key_obj = serialization.load_pem_private_key(
            private_key_pem.encode(),
            password=None,
            backend=default_backend()
        )
        
        conn_params = {
            "account": self.config.get_snowflake_account(),
            "user": self.config.get_snowflake_user(),
            "role": "SNOWFLAKE_INTELLIGENCE_ADMIN",
            "warehouse": self.config.get_warehouse(),
            "database": self.config.get_database(),
            "schema": self.config.get_schema(),
            "private_key": private_key_obj,
        }
        
        max_id = 1
        try:
            conn = snowflake.connector.connect(**conn_params)
            cursor = conn.cursor()
            cursor.execute(
                f"SELECT MAX(CUSTOMER_ID) as MAX_ID FROM "
                f"{self.config.get_database()}.RAW.CUSTOMERS"
            )
            result = cursor.fetchone()
            if result and result[0]:
                max_id = result[0]
            logger.info(f"Max customer ID in database: {max_id}")
            cursor.close()
            conn.close()
        except Exception as e:
            logger.error(f"Error fetching max customer ID: {e}")
            raise
        
        return max_id

    def insert_order(self, order: Order) -> None:
        row = order.to_dict()
        offset_token = f"order_{order.order_id}"
        
        self.orders_channel.append_row(row, offset_token)
        logger.debug(f"Order {order.order_id} inserted with offset {offset_token}")

    def insert_orders(self, orders: List[Order]) -> None:
        if not orders:
            return
        
        start_offset = f"order_{orders[0].order_id}"
        end_offset = f"order_{orders[-1].order_id}"
        
        rows = [order.to_dict() for order in orders]
        
        self.orders_channel.append_rows(rows, start_offset, end_offset)
        logger.debug(
            f"Inserted {len(orders)} orders (offset range: {start_offset} to {end_offset})"
        )

    def insert_order_items(self, items: List[OrderItem]) -> None:
        if not items:
            return
        
        start_offset = f"item_{items[0].order_item_id}"
        end_offset = f"item_{items[-1].order_item_id}"
        
        rows = [item.to_dict() for item in items]
        
        self.order_items_channel.append_rows(rows, start_offset, end_offset)
        logger.debug(
            f"Inserted {len(items)} order items (offset range: {start_offset} to {end_offset})"
        )

    def get_latest_order_offset(self) -> Optional[str]:
        return self.orders_channel.get_latest_committed_offset_token()

    def get_latest_order_item_offset(self) -> Optional[str]:
        return self.order_items_channel.get_latest_committed_offset_token()

    def close(self) -> None:
        try:
            logger.info("Closing channels...")
            if hasattr(self, "orders_channel"):
                self.orders_channel.close()
            if hasattr(self, "order_items_channel"):
                self.order_items_channel.close()
            
            logger.info("Closing clients...")
            if hasattr(self, "orders_client"):
                self.orders_client.close()
            if hasattr(self, "order_items_client"):
                self.order_items_client.close()
            
            logger.info("Snowpipe Streaming manager closed successfully")
        except Exception as e:
            logger.error(f"Error closing Snowpipe Streaming manager: {e}")

File: snowpipe-streaming-java/src/main/java/com/snowflake/demo/SnowpipeStreamingManager.java
package com.snowflake.demo;

import com.snowflake.ingest.streaming.SnowflakeStreamingIngestChannel;
import com.snowflake.ingest.streaming.SnowflakeStreamingIngestClient;
import com.snowflake.ingest.streaming.SnowflakeStreamingIngestClientFactory;
import com.snowflake.ingest.streaming.OpenChannelResult;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.security.KeyFactory;
import java.security.PrivateKey;
import java.security.spec.PKCS8EncodedKeySpec;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.stream.Collectors;

public class SnowpipeStreamingManager {
    private static final Logger logger = LoggerFactory.getLogger(SnowpipeStreamingManager.class);

    private final SnowflakeStreamingIngestClient ordersClient;
    private final SnowflakeStreamingIngestClient orderItemsClient;
    private final SnowflakeStreamingIngestChannel ordersChannel;
    private final SnowflakeStreamingIngestChannel orderItemsChannel;
    private final ConfigManager config;
    private final Properties connectionProps;

    public SnowpipeStreamingManager(ConfigManager config) throws Exception {
        this(config, -1);
    }

    public SnowpipeStreamingManager(ConfigManager config, int instanceId) throws Exception {
        this.config = config;

        this.connectionProps = new Properties();
        connectionProps.put("account", config.getSnowflakeAccount());
        connectionProps.put("user", config.getSnowflakeUser());
        connectionProps.put("url", config.getSnowflakeUrl());
        connectionProps.put("private_key", config.getPrivateKey());
        connectionProps.put("role", "SNOWFLAKE_INTELLIGENCE_ADMIN");
        connectionProps.put("warehouse", config.getWarehouse());
        connectionProps.put("db", config.getDatabase());
        connectionProps.put("schema", config.getSchema());

        String channelSuffix = (instanceId >= 0) ? "_instance_" + instanceId : "";
        logger.info("Creating Snowflake Streaming clients and opening channels{}", 
                   instanceId >= 0 ? " for instance " + instanceId : "...");
        
        this.ordersClient = SnowflakeStreamingIngestClientFactory.builder(
                "ORDERS_CLIENT_" + java.util.UUID.randomUUID(),
                config.getDatabase(),
                config.getSchema(),
                config.getProperty("pipe.orders.name")
        )
                .setProperties(connectionProps)
                .build();
        
        this.orderItemsClient = SnowflakeStreamingIngestClientFactory.builder(
                "ORDER_ITEMS_CLIENT_" + java.util.UUID.randomUUID(),
                config.getDatabase(),
                config.getSchema(),
                config.getProperty("pipe.order_items.name")
        )
                .setProperties(connectionProps)
                .build();

        this.ordersChannel = openChannel(
                ordersClient,
                config.getProperty("channel.orders.name") + channelSuffix
        );

        this.orderItemsChannel = openChannel(
                orderItemsClient,
                config.getProperty("channel.order_items.name") + channelSuffix
        );

        logger.info("All clients and channels initialized successfully");
    }

    private SnowflakeStreamingIngestChannel openChannel(SnowflakeStreamingIngestClient client, String channelName) throws Exception {
        String initialOffset = "0";
        OpenChannelResult result = client.openChannel(channelName, initialOffset);
        SnowflakeStreamingIngestChannel channel = result.getChannel();
        
        String latestOffset = channel.getLatestCommittedOffsetToken();
        logger.info("Channel {} opened. Latest committed offset: {}", channelName, 
                    latestOffset != null ? latestOffset : "NULL (new channel)");
        
        return channel;
    }

    private PrivateKey parsePrivateKey(String privateKeyPEM) throws Exception {
        String privateKeyPEMClean = privateKeyPEM
            .replace("-----BEGIN PRIVATE KEY-----", "")
            .replace("-----END PRIVATE KEY-----", "")
            .replaceAll("\\s", "");
        
        byte[] encoded = Base64.getDecoder().decode(privateKeyPEMClean);
        PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(encoded);
        KeyFactory kf = KeyFactory.getInstance("RSA");
        return kf.generatePrivate(keySpec);
    }

    public int getMaxCustomerId() throws Exception {
        String jdbcUrl = config.getSnowflakeUrl().replace(":443", "").replace("https://", "jdbc:snowflake://");
        
        Properties jdbcProps = new Properties();
        jdbcProps.put("account", config.getSnowflakeAccount());
        jdbcProps.put("user", config.getSnowflakeUser());
        jdbcProps.put("role", "SNOWFLAKE_INTELLIGENCE_ADMIN");
        jdbcProps.put("warehouse", config.getWarehouse());
        jdbcProps.put("db", config.getDatabase());
        jdbcProps.put("schema", config.getSchema());
        
        PrivateKey privateKey = parsePrivateKey(config.getPrivateKey());
        jdbcProps.put("privateKey", privateKey);
        
        int maxId = 1;
        try (Connection conn = DriverManager.getConnection(jdbcUrl, jdbcProps);
             Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery("SELECT MAX(CUSTOMER_ID) as MAX_ID FROM " + 
                                            config.getDatabase() + "." + 
                                            config.getSchema() + ".CUSTOMERS")) {
            if (rs.next()) {
                maxId = rs.getInt("MAX_ID");
            }
            logger.info("Max customer ID in database: {}", maxId);
        }
        
        return maxId;
    }

    public void insertOrder(Order order) throws Exception {
        Map<String, Object> row = order.toMap();
        String offsetToken = "order_" + order.getOrderId();
        
        ordersChannel.appendRow(row, offsetToken);
        logger.debug("Order {} inserted with offset {}", order.getOrderId(), offsetToken);
    }

    public void insertOrders(List<Order> orders) throws Exception {
        if (orders.isEmpty()) {
            return;
        }
        
        String startOffset = "order_" + orders.get(0).getOrderId();
        String endOffset = "order_" + orders.get(orders.size() - 1).getOrderId();
        
        List<Map<String, Object>> rows = orders.stream()
                .map(Order::toMap)
                .collect(Collectors.toList());
        
        ordersChannel.appendRows(rows, startOffset, endOffset);
        logger.debug("Inserted {} orders (offset range: {} to {})", 
                    orders.size(), startOffset, endOffset);
    }

    public void insertOrderItems(List<OrderItem> items) throws Exception {
        if (items.isEmpty()) {
            return;
        }
        
        String startOffset = "item_" + items.get(0).getOrderItemId();
        String endOffset = "item_" + items.get(items.size() - 1).getOrderItemId();
        
        List<Map<String, Object>> rows = items.stream()
                .map(OrderItem::toMap)
                .collect(Collectors.toList());
        
        orderItemsChannel.appendRows(rows, startOffset, endOffset);
        logger.debug("Inserted {} order items (offset range: {} to {})", 
                    items.size(), startOffset, endOffset);
    }

    public String getLatestOrderOffset() {
        return ordersChannel.getLatestCommittedOffsetToken();
    }

    public String getLatestOrderItemOffset() {
        return orderItemsChannel.getLatestCommittedOffsetToken();
    }

    public void close() {
        try {
            logger.info("Closing channels...");
            if (ordersChannel != null) ordersChannel.close();
            if (orderItemsChannel != null) orderItemsChannel.close();
            
            logger.info("Closing clients...");
            if (ordersClient != null) ordersClient.close();
            if (orderItemsClient != null) orderItemsClient.close();
            
            logger.info("Snowpipe Streaming manager closed successfully");
        } catch (Exception e) {
            logger.error("Error closing Snowpipe Streaming manager", e);
        }
    }
}
